### ISSUE TITLE
live.text / live.tab: Max hangs when reloading patcher

### ISSUE THREAD

**Issue by benbrackenc74** [Cycling74/miraweb#58]
Max will hang when you re-open a patcher containing a mira.frame object that you had previously connected to via the browser. It seems to have to do with the number of live.text objects.

Steps to repro:
1. In Max, open patcher
2. Open browser to connect to mira.frame
3. Back in Max, close the patcher
4. In Max, re-open the patcher - Max hangs

Max 7.3.2, latest miraweb package (1.2.1), FF 51.0.1 (64-bit), can repro in Chrome and Safari as well.

[add_params.maxpat.zip](https://github.com/Cycling74/miraweb/files/802537/add_params.maxpat.zip)

**Comment by benbrackenc74** [comment]
This seems to have something to do with utilizing SVGs.

**Comment by 1offby** [comment]
Can reproduce. Add_params.maxpat.zip yields a hang log similar to others I've been experiencing in similar use cases.

**Comment by 1offby** [comment]
Any progress here?
I'm about to complete a few things that use this feature (live.xxx and .svg) and in good position to test any miraweb versions that may resolve this bug.

**Comment by benbrackenc74** [comment]
No new work has been done here. happens with fpic if you have enough of them. There is no caching of reused sprites, so this is maybe an area of optimization that could happen. Maybe something more needs to happen in destroy() to fully clear out the sprite? Unsure ATM how to solve.

**Comment by 1offby** [comment]
Here's an actual Max crash prompted by opening a miraweb socket on Chrome on the local machine
[mirawebcrash.txt](https://github.com/Cycling74/miraweb/files/845412/mirawebcrash.txt)

**Comment by 1offby** [comment]
For what it's worth I have been able to get a similar hang using the Mira App.
Definitely do want to use MiraWeb in situations where reliability is an important consideration; hope it's possible to avoid recurring Max application hangs/crashes.
Many thanks for making a great Max Package!

**Comment by starakaj** [comment]
Hmm. Memory leaks? Race conditions? Hard to say---let's find out.

**Comment by starakaj** [comment]
Can reproduce

**Comment by starakaj** [comment]
Ok, you don't have to re-open the patch---you just have to open a patch containing a lot of SVGs while Max is already connected to Miraweb (or mira)

**Comment by 1offby** [comment]
In my experience the hang can also start when changing the .svg contained in a bunch of live.text objects within the mira.frame via the "pictures" message.
Thank you for looking into this!

**Comment by starakaj** [comment]
Hmmmmmmmmmmmm that's interesting. So maybe it has nothing to do with creating a brand new patch...

**Comment by benbrackenc74** [comment]
I think that making a new patch, or reloading the previous one, is just an
easy way to make it happen. It seems to me like a memory thing, or like
textures/sprites aren't getting destroyed/freed and start stacking up.

**Comment by starakaj** [comment]
Just pausing execution on the application when it hangs seems to suggest this might be related to JUCE's MessageManagerLock

**Comment by starakaj** [comment]
K, I'm pretty sure now that this has to do with mixing-and-matching different kinds of critical sections, which is causing a deadlock. It's not _immediately_ obvious how to solve this, but I can try a couple things.

**Comment by 1offby** [comment]
FYI the MiraWeb implementation where I get the most hangs includes conditions where a lot of live.text objects get pictures or set messages almost simultaneously. Seems like a situation where resources could get mixed up.

**Comment by 1offby** [comment]
Well, this was an exciting series-of-comments-on-an-issue.
Any progress?

**Comment by starakaj** [comment]
@1offby I banged on it for a few hours on Friday---unfortunately it's a bit sticky and involves a deadlock among pieces that I don't totally understand. I'm trying to figure out the best way to solve it without touching too much code. It's being looked at, don't worry!

**Comment by 1offby** [comment]
I appreciate the update, and the effort. Many thanks!
I checked the RAM impact of lots of live.text objects compared to a similar number of fpics,
those live.text thingys use a lot of resources, and I thought that might contribute to the issue. I am changing the way I build UI for Miraweb to use fpic wherever possible, and will report whether that resolves the hang...
Nonetheless looking forward to conditions where Max doesn't hang even when there are lots of live.xxx with svg in mira.frame.

**Comment by 1offby** [comment]
I took one of the patchers I'm working on that had a lot of .svg being swapped in and out of live.text and replaced those live.text with fpic.
Similar hangs persist.
[hangs.zip](https://github.com/Cycling74/miraweb/files/856782/hangs.zip)
Interestingly the memory impact in terms of RAM use, which I thought would go down with fpic replacing live.text, remained the same.

**Comment by benbrackenc74** [comment]
@1offby yes, fpic has the same problem, as mentioned above.

**Comment by 1offby** [comment]
Darn.

**Comment by 1offby** [comment]
Checking in.
I have client apps in beta but app hanging is a no-go.
Does this bug look solve-able?
Thank you for any update

**Comment by benbrackenc74** [comment]
@1offby this ticket will be touched when we have a fix. We do not have a timetable for this.

**Comment by 1offby** [comment]
For what it's worth this hang can also happen when closing a max patch that is connected to Mira app or Miraweb.

**Comment by 1offby** [comment]
@fde31 What does the Component: Max Objects label mean?
I'm guessing this indicates that the issue relates to Max objects, but which ones? Mira.frame or live.text/fpic?

**Comment by fde31** [comment]
@1offby sorry for the confusion. We renamed it to "Backend" and just use it as a generic label for things that need to be tackled on the Max/backend side compared to ones on the client/frontend libs.

**Comment by starakaj** [comment]
Alright, let's see if I can't make some progress here...

**Comment by starakaj** [comment]
Okay @benbrackenc74, would you mind giving this a test? You'll need to pull the latest max and rebuild mira.frame. Looks like there are certain functions that libXebra expects to be executed from the main thread that were not, in fact, being executed from the main thread. With this fixed, this problem seems to go away.

**Comment by 1offby** [comment]
Thank you!
I am excited.

**Comment by benbrackenc74** [comment]
I can confirm that this will be fixed for Max 7.3.4. Well done @starakaj.

**Comment by 1offby** [comment]
Thank you very much for your work on this!

**Comment by benbrackenc74** [comment]
There are still some issues with this work. @1offby says:

Here is a case where multiple tabs within sub-patchers in presentation mode render images unpredictably. Also unpredictable when all are not in presentation mode, but this example is more "real world" in my case.

The attached patcher works pretty well in the Mira app, and OK when the browser is on the local machine, but not great/or at all when the connected browser (chrome or firefox) is on an iPad.

[NewMirawebReload.maxpat.zip](https://github.com/Cycling74/miraweb/files/1027404/NewMirawebReload.maxpat.zip)

**Comment by starakaj** [comment]
Well at least it's not freezing! Taking a look

**Comment by starakaj** [comment]
This is working pretty okay for me, it's just really slow. So to help me debug a bit, @1offby, what do you mean when you say it "renders images unpredictably?"

**Comment by 1offby** [comment]
[https://youtu.be/d74R19uJMKE](https://youtu.be/d74R19uJMKE)
In the example patch, choosing a tab sometimes shows blank buttons rather than buttons with images. A tab that showed images when selected once will be blank when selected again after switching to another tab. These images will either take >10 sec to load or not load at all until the iPad screen is tapped. This is what I'm calling "unexpected," because I expect selecting a tab to load/refresh all that tab's images relatively quickly (<5 sec? Faster?). To me it's also unexpected that a correctly drawn tab will no longer be correct when it's re-selected. I also do not expect tapping the iPad screen to assist with drawing the UI elements--I expect all elements to be drawn independent of my interaction.

In my actual use case this situation is exacerbated, maybe because the GUI isn't just dummy live.text buttons but buttons that interact with each other with images that are being loaded dynamically.

Yes, Max not hanging is a HUGE improvement: thank you very much!

**Comment by benbrackenc74** [comment]
Maybe this is turning into a feature request to somehow cache resources that are used multiple times?

**Comment by 1offby** [comment]
Yeah, whatever this is it ain't #58 as originally described.
Is the ability of MiraWeb to (at least kinda)quickly draw UI elements on tab switch, including images in live.text, a new "feature?" I have assumed this was expected functionality.

**Comment by fde31** [comment]
I agree that this is neither ideal nor expected behaviour. We are investigating and figuring out a solution. Just to give a bit of insight: we are trying to identify if the work on this should be done in MiraWeb or in Xebra.js. The latter would make it more widely usable but maybe also reduce flexibility for users of the lib version. So yeah, we'll keep you updated and thanks for the ongoing feedback and contributions to MiraWeb, we highly appreciate it and are eager to make MW better with the help from the community.

**Comment by 1offby** [comment]
I enjoy being part of the process!
There's a lot of value to me and my clients in MiraWeb/Xebra.js.
GUI-in-the-browser is a real step forward for a lot of applications; I appreciate you and your colleauges' work on improving its real-world implementation!

**Comment by 1offby** [comment]
Regarding Ben's comment above, if there was a method similar to preload messages to sfplay~ that held image assets in some kind of buffer or cache for quicker recall in the browser, I'd use it.

**Comment by starakaj** [comment]
Need to follow up on this one.
